tests/tests/integration_tests.rs

```diff
@@ -295,3 +294,45 @@ async fn ws_close_pending_subscription_when_server_terminated() {
 
 	panic!("subscription keeps sending messages after server shutdown");
 }
+
+#[tokio::test]
+async fn ws_server_should_stop_subscription_after_client_drop() {
+	use futures::{channel::mpsc, SinkExt, StreamExt};
+	use jsonrpsee::{ws_server::WsServerBuilder, RpcModule};
+
+	let server = WsServerBuilder::default().build("127.0.0.1:0").await.unwrap();
+	let server_url = format!("ws://{}", server.local_addr().unwrap());
+
+	let (tx, mut rx) = mpsc::channel(1);
+	let mut module = RpcModule::new(tx);
+
+	module
+		.register_subscription("subscribe_hello", "unsubscribe_hello", |_, mut sink, mut tx| {
+			tokio::spawn(async move {
+				let close_err = loop {
+					if let Err(Error::SubscriptionClosed(err)) = sink.send(&1) {
+						break err;
+					}
+					tokio::time::sleep(Duration::from_millis(100)).await;
+				};
+				let send_back = Arc::make_mut(&mut tx);
+				send_back.feed(close_err).await.unwrap();
+			});
+			Ok(())
+		})
+		.unwrap();
+
+	tokio::spawn(async move { server.start(module).await });
+
+	let client = WsClientBuilder::default().build(&server_url).await.unwrap();
+
+	let mut sub: Subscription<usize> =
+		client.subscribe("subscribe_hello", ParamsSer::NoParams, "unsubscribe_hello").await.unwrap();
+
+	let res = sub.next().await.unwrap();
+
+	assert_eq!(res.as_ref(), Some(&1));
+	drop(client);
+	// assert that the server received `SubscriptionClosed` after the client was dropped.
+	assert!(rx.next().await.is_some());
```

Comment:
This is a bit opaque unless one knows exactly what is going on.

Maybe we can do something like this?

```
	let sub_closed_err = rx.next().await.expect("Server received `SubscriptionClosed` after the client was dropped.");
	assert!(format!("{:?}", sub_closed_err).contains("is closed but not yet dropped"));
```

It's a bit annoying that `SubscriptionClosedErr` has a private member (and that there's no way for the client to know what the subscription ID is) or we could have used `assert_matches` here for better readability...

Reply:
Yepp, I think we could replace SubscriptionClosed(String) with a better type.

Perhaps with a `kind enum`, to distinguish why the subscription was closed the error message is quite useful atm